Feature/http parser joushin

Hawardshin opened the pull request:
## ✅ 풀\_리퀘스트 체크리스트

<!--
하나씩 확인 후 체크박스에 표시해주세요.
-->

- [x] PR 제목: [Feature/Fix/Refactor...] 작업 내용 한 줄 요약 (브랜치 이름) #이슈번호
- [x] commit message 가 적절한지 확인해주세요.
- [x] 적절한 branch 로 요청했는지 확인해주세요.
- [x] Assignees, Label 을 붙여주세요.
- [x] 주의 사항과 관련해 꼭 확인해야 할 사람이 있다면 Reviewer 로 등록해주세요.
- [x] PR이 승인된 경우 해당 브랜치는 삭제해 주세요!

<br/>

## 🔄 변경 사항

<!-- 해당 pr에서 작업한 내역을 적어주세요. 처음엔 간단하게 요약, list 형식으로 세부사항 작성 -->
Http Request 파싱을 완료했습니다.

- 이전에 만들었던 body parsing과 @besyia0k0 콩의 header parsing code를 합쳤습니다.
- ~~코드 진행 중 같은 클라이언트로 부터 정상적인 요청을 지속적으로 받게 될 때 이전의 결과값이 영향을 주는 것을 확인했습니다.~~
- ~~그에 따라서 이전의 header결과와 finish로 끝나는 경우 전부 다시 first로 상태를 변경해줬습니다.~~
- content-length가 존재하거나 chunked인 경우에는 전부 body로 status를 변경해줬습니다.
- Location block에서 wild카드에 해당하는 규칙을 추가했습니다.
  - 예를 들어 `location /.html{ ....}` 로 들어오는 경우 모든 path가 .html로 끝나는 파일은 저 block을 반환합니다.
- 그 외에 server Manager의 parsing 부분에서 함수 호출 순서와 로직 추가가 있습니다.
- 추가적으로 Tools 폴더를 밖으로 뺐습니다.
- 추가적으로 테스터의 폴더 구조를 위한 파일을 조금 추가했습니다.

++ file read write이벤트를 등록하는 코드를 추가했습니다.(Server Manager)
- 파일 Read event 와 write이벤트를 등록하는 함수를 만들었습니다.
- 파일이 Read가 끝나면 자동으로 client Write event에 등록을 해주는 식으로 동작합니다.
- Udata에 status 와 file을 읽고 쓸 때 사용할 버퍼가 추가되었습니다.
<br/>

## 📎 변경한 이유

- location block의 와일드 카드 구현이 서브젝트에 요구사항이여서 추가했습니다.
- Http body parsing을 test하던 중 문제를 발견해서 그 부분들을 수정했습니다.

++
- 추가적으로 Post요청을 처리하거나 Error page를 띄우기 위해서는 file을 오픈하는게 먼저라고 생각해서 file open에 관련해서 event를 추가하고 udata를 변경했습니다.

<br/>

<!-- 관련되어있는 Issue Number 를 작성하세요! 해당 이슈를 이곳에 적으면 pr merge 이후 해당 이슈는 자동으로 close 됩니다. -->

closes: #Issue Number

## 👨🏻‍💻 테스트 체크리스트

<!-- 테스트 사항이 있다면 작성해 주세요! -->

- main문에 적어둔 대로 body에 대해서 chunked와 일반일 때 parsing하는 코드를 테스트 했습니다.
- 그리고 location 블록 와일드 카드도 제대로 작동하는지 확인했습니다.
- 이후 전체 코드 make해서 curl명령어로 명령 보냈을 때 제대로 body 출력하는지 확인했습니다. (header는 콩이 해봐서 따로 안 했습니다.)
- 같은 클라이언트가 정상적인 요청을 여러번 보낼 때에 대해서 테스트를 완료해서 제대로 body를 parsing하는 것 확인 했습니다. 
<br/>

## 📌 변경 및 주의 사항
<!--
변경사항 및 주의 사항이 있다면 적어주세요.
주의 사항과 관련해 꼭 확인해야 할 사람이 있다면 리뷰어로 등록해주세요. (다른 사람이 작성한 코드 수정 등)
코드 리뷰 시 더 꼼꼼하게 확인 받고 싶은 부분이 있다면 적어주세요.
-->
- 콩이 작성한 코드에 제가 추가한 body_.clear()와 header_.clear()가 있습니다. 혹시나 더 지워야 하는게 있는지 또는 제가 잘못 수정한 거 일수도 있으니까 한번 확인 부탁해요~~
<br/>

Hawardshin committed:
Feat: Http Request chunked incoding
- 청크된 인코딩 받는 함수 추가합니다.

Hawardshin committed:
Merge branch 'develop' of github.com:42masterplan/webserv into feature/HttpParser_joushin

Hawardshin committed:
Feat: Add Conf file wild Card
- 이제 conf파일에서 . 으로 끝나는 wild카드를 받아서 사용합니다.

Hawardshin committed:
Feat: Http request 처리 로직 업데이트

Hawardshin committed:
Feat: Combine body and header parsing

Hawardshin committed:
Feat: Add HTTP Body Parsing
- HTTP Body parsing 함수를 작성했습니다.
- HTTP 메세지 파싱을 마쳤습니다.
- Test용 main문도 함께 변경사항과 makefile로 테스트 가능하게 함께 올리겠습니다.

Hawardshin committed:
Test: Add Body parsing Test

Hawardshin committed:
Fix: Fix lots of Request Same client Parse
- 한명의 클라이언트에게 여러번의 요청이 오는 경우에 문제가 생겼습니다.
- 그에 따라 body와 header를 비워줬습니다.
- 그 외 main문으로 기초적인 chunked와 content-length 기준으로 들어오는 메시지를 잘 받는걸 확인 했습니다.

Hawardshin committed:
Merge branch 'develop' of github.com:42masterplan/webserv into feature/HttpParser_joushin

Hawardshin committed:
Test: tester file add
- 테스트 할때 사용 하는 파일들을 추가했습니다.

Hawardshin committed:
Rename: Tools folder move
- Tools 폴더를 밖으로 뺏습니다.

Hawardshin committed:
Feat: Add File Read Write Event And Edit Udata class for file Read write
- 파일을 read와 write 할 때도 이벤트를 등록해서 해야하기 때문에 그에
  맞는 코드들을 추가했습니다.
- 아직 테스트는 많이 못 해봐서 코드 짜면서 같이 테스트 해봐야 할 것
  같습니다.

Hawardshin committed:
Add gitignore

Hawardshin committed:
Fix: static to member variable
-Body parsing 할 때 동시에 여러개를 하면 static 변수상 문제가 생길 것
같아 각 객체마다 가질 수 있도록 추가해줬습니다.

Hawardshin committed:
Fix: Use once for this class
- 우리는 해당 클래스의 객체를 finish가 된 이후에 다시 사용할 일이 없기
  때문에 수정했던 부분을 삭제합니다.

besyia0k0 commented on modules/UData/HttpRequest/HttpRequest.cpp:
```diff
@@ -209,9 +232,60 @@ void	HttpRequest::checkHeader(void) {
+		if (to_read_ == 0 && raw_data.size() >= 2){
+			// std::cout << "end"<<std::endl;
+			read_state_ = false;
+			getLine(raw_data);
```
chunked일 때 들어온 length와 실제 들어온 값의 길이가 다르면 오류 처리를 해야하나용?

Hawardshin replied:
앗 처리하겠습니당

besyia0k0 commented on modules/UData/HttpRequest/HttpRequest.cpp:
```diff
@@ -209,9 +232,60 @@ void	HttpRequest::checkHeader(void) {
+		std::copy(raw_data.begin(), raw_data.begin() + content_length_,  std::back_inserter(body_));
+		raw_data.erase(raw_data.begin(),raw_data.begin() + content_length_);
+		parse_status_ = FINISH;
+		if (raw_data.size() == 0)
```
뒤에 응답이 이미 와있을수도 있는데 raw_data size가 0인 것은 확인하지 않고 content-length가 넘으면 복사하고 true 반납해도 되지 않을까요?

Hawardshin replied:
앗 맞아요!! 제가 한 객체를 여러번 쓰려다 보니 에러가 있었네요

besyia0k0 commented on modules/webserv/ServManager.cpp:
```diff
@@ -166,11 +175,25 @@ void  ServManager::sockReadable(struct kevent *cur_event){
 		buff_[rlen] = '\0';
 		raw_data_ref.insert(raw_data_ref.end(), buff_, buff_ + rlen);
-		if (cur_udata->http_request_.size() == 0){
+		if (cur_udata->http_request_.size() == 0 || \
```
이걸 계속 돌려야 할 것 같아요 raw_data에서 최대한 파싱할 수 있는 대로 하는 방법으로 ..? 이건 나중에 제가 수정해볼게용

Hawardshin replied:
아항 여러개의 request가 한번에 들어오는 경우군요 !!

besyia0k0 commented on modules/webserv/ServManager.cpp:
```diff
@@ -166,11 +175,25 @@ void  ServManager::sockReadable(struct kevent *cur_event){
-		cur_udata->http_request_[cur_udata->http_request_.size() - 1].parse(raw_data_ref);
+		cur_udata->http_request_.back().parse(raw_data_ref);
+		cur_udata->http_request_.back().printBodyInfo();
+		if (cur_udata->http_request_.back().getRequestError()){//아에 잘못 된 형식으로 메세지가 온 경우들에 대해서 Request 단에서 에러를 처리해줍니다.
```
이거 response만드는데에서 하는거 찬성임니당

besyia0k0 reviewed (approved):
LGTM 👍

Hawardshin committed:
Fix : Chunked line Error Add
- 청크드 된 경우 에러로 감지하는 코드를 추가했습니다.

Hawardshin committed:
Fix: change detail for http parsing
- http body parsing의 디테일한 부분을 삭제했습니다.

Hawardshin committed:
Fix: parsebdy edit
- 바디 파싱하는 부분을 수정했습니다
- 테스트 할 때 하나의 클라이언트가 여러번의 요청을 보내는 경우에 대해서
  하나의 객체로 처리하려고 하다 보니 이런 상황이 생겼네요~~

besyia0k0 commented on modules/UData/HttpRequest/HttpRequest.cpp:
```diff
@@ -209,9 +232,60 @@ void	HttpRequest::checkHeader(void) {
+ * 2.이후 16진수를 해석해서 그만큼 또 raw data를 비워줍니다.
+ * 3. 0이 나올 때까지 반복합니다.
+ * @param raw_data 들어온 데이터를 저장하는 저장소
+ * @return true 파싱이 끝남
```
버퍼를 더 받아야 하는 경우: false, 에러든 뭐든 아무튼 작동했다: true

Hawardshin committed:
Fix: return state
- parseBody 함수에서 return state를 변경했습니다

Hawardshin committed:
Fix: parse body refactor

Hawardshin committed:
Fix: body parsing for retrun true
- 청크드에서 raw_data를 읽었을 때 body 를 다시 받는 상황이 생겨서
  수정했습니다~
